[ML] Single Metric Viewer embeddable: ensures chart height is responsive on resize

alvarezmelissa87 opened the pull request:
## Summary

This PR ensures the SMV chart in dashboards is responsive.

This PR takes the height on panel change and calculates the percentage height of each part of the SMV chart - these heights are then passed to the chart element.

<img width="1332" alt="image" src="https://github.com/elastic/kibana/assets/6446462/fadcc0f5-420a-4417-872e-bec3fb9f14b6">




### Checklist

Delete any items that are not applicable to this PR.

- [ ] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/main/packages/kbn-i18n/README.md)
- [ ] [Documentation](https://www.elastic.co/guide/en/kibana/master/development-documentation.html) was added for features that require explanation or tutorials
- [ ] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios
- [ ] [Flaky Test Runner](https://ci-stats.kibana.dev/trigger_flaky_test_runner/1) was used on any tests changed
- [ ] Any UI touched in this PR is usable by keyboard only (learn more about [keyboard accessibility](https://webaim.org/techniques/keyboard/))
- [ ] Any UI touched in this PR does not create any new axe failures (run axe in browser: [FF](https://addons.mozilla.org/en-US/firefox/addon/axe-devtools/), [Chrome](https://chrome.google.com/webstore/detail/axe-web-accessibility-tes/lhdoppojpmngadmnindnejefpokejbdd?hl=en-US))
- [ ] If a plugin configuration key changed, check if it needs to be allowlisted in the cloud and added to the [docker list](https://github.com/elastic/kibana/blob/main/src/dev/build/tasks/os_packages/docker_generator/resources/base/bin/kibana-docker)
- [ ] This renders correctly on smaller devices using a responsive layout. (You can test this [in your browser](https://www.browserstack.com/guide/responsive-testing-on-local-server))
- [ ] This was checked for [cross-browser compatibility](https://www.elastic.co/support/matrix#matrix_browsers)

walterra commented:
I created a PR against yours with some fixes: https://github.com/alvarezmelissa87/kibana/pull/5

- marginTop/marginBottom/chartSpacing should not be based on percentages, this messes with the offsets and caused a gap between focus chart and context chart.
- To have one source of truth I updated `getSvgHeight` to consider `heights` from props. This allowed to not use the `svgHeight` passed in via props.
- The height coming from the embedded panel to be passed on to the time series chart includes the height taken by the title (32px), I subtracted that again when passed in to `getChartHeights`.

elasticmachine commented:
Pinging @elastic/ml-ui (:ml)

peteharverson commented:
Done first round of testing - overall a big improvement. Some items I've noticed:

- If the panel resize is purely vertical on height, the content doesn't resize - this only seems to happen when you resize width too.
- The chart can overflow into the zoom tools:

<img width="463" alt="Screenshot 2024-06-19 at 13 13 34" src="https://github.com/elastic/kibana/assets/7405507/46188c4a-d9b1-4123-9e2b-b33a690f67b4">

compared to inside the ML page:

<img width="462" alt="Screenshot 2024-06-19 at 13 14 41" src="https://github.com/elastic/kibana/assets/7405507/935a7810-9626-48d0-ac47-ea7f65e9f63a">

walterra commented on x-pack/plugins/ml/public/shared_components/single_metric_viewer/single_metric_viewer.tsx:
```diff
@@ -150,11 +153,11 @@ const SingleMetricViewerWrapper: FC<SingleMetricViewerPropsWithDeps> = ({
     throttle((e: { width: number; height: number }) => {
-      if (Math.abs(chartWidth - e.width) > minElemAndChartDiff) {
-        setChartWidth(e.width);
+      if (Math.abs(chartDimensions.width - e.width) > minElemAndChartDiff) {
```
I suspect this needs to be updated to check the height too, this is what might be causing the chart height not being propagated when you change the height of a panel only.

alvarezmelissa87 replied:
Yep, good spot - updated in a5474353d89360f539a1f68abd0ded0be695f08f

peteharverson commented on x-pack/plugins/ml/public/application/timeseriesexplorer/components/timeseries_chart/timeseries_chart_with_tooltip.tsx:
```diff
@@ -21,6 +21,44 @@ import { getControlsForDetector } from '../../get_controls_for_detector';
 
+const percentFocusZoomPanelHeight = 0.051;
+const percentFocusChartHeight = 0.634;
+const percentContextChartHeight = 0.122;
```
@alvarezmelissa87 @walterra what do you think about giving the context chart a fixed height, and have the focus chart fill the rest of the available height?

peteharverson replied:
Fixed height for the context chart is working well!

alvarezmelissa87 commented:
@elasticmachine merge upstream

alvarezmelissa87 commented:
@pete - thanks for giving this a test! 
Updated here a5474353d89360f539a1f68abd0ded0be695f08f to ensure the resize happens when the panel is resized vertically only.

peteharverson commented:
Gave this another test against https://github.com/elastic/kibana/pull/185907/commits/a5474353d89360f539a1f68abd0ded0be695f08f. Vertical-only resize and reduced min height of the focus chart look good.

Any thoughts on:
- Focus chart overflowing into the zoom tools panel
- Fixing the height of the context chart
-

alvarezmelissa87 commented:
@elasticmachine merge upstream

alvarezmelissa87 commented:
@elasticmachine merge upstream

alvarezmelissa87 commented:
@elasticmachine merge upstream

alvarezmelissa87 commented:
This has been updated and is ready for a final look when you get a chance 🙏 cc @walterra, @peteharverson 

The context chart and zoom panel are now fixed heights and the focus chart is the only thing that resizes. Also fixed the issue of the focus chart overflow.

walterra commented on x-pack/plugins/ml/public/application/timeseriesexplorer/components/timeseries_chart/timeseries_chart.js:
```diff
@@ -90,10 +93,27 @@ const anomalyGrayScale = d3.scale
 
-function getSvgHeight(showAnnotations) {
+function getChartHeights(height) {
+  const actualHeight = height < minSvgHeight ? minSvgHeight : height;
```
Nit: Could be `const actualHeight = Math.max(height, minSvgHeight);`.

walterra commented on x-pack/plugins/ml/public/application/timeseriesexplorer/components/timeseries_chart/timeseries_chart.js:
```diff
@@ -90,10 +93,27 @@ const anomalyGrayScale = d3.scale
+
+function getSvgHeight(showAnnotations, incomingHeight) {
   const adjustedAnnotationHeight = showAnnotations ? annotationHeight : 0;
+  const incomingHeightActual =
```
Nit: Could be `const incomingHeightActual = Math.max(incomingHeight || 0, minSvgHeight);`.

walterra commented:
Great update, looks like the previous issues are resolved! A thing I'm noticing is that the bottom margin in the panel is not consistent, for example here it's quite high, but that's not always the case:

<img width="767" alt="image" src="https://github.com/elastic/kibana/assets/230104/f29f579d-53ce-480c-9ac4-1e14131f9ea4">

peteharverson commented:
> A thing I'm noticing is that the bottom margin in the panel is not consistent, for example here it's quite high, but that's not always the case

Is the bottom margin getting a % of the available height? Not sure if it's related, but at lower heights, I see a scrollbar even thought all the chart is (just) visible:

<img width="368" alt="Screenshot 2024-07-02 at 11 35 03" src="https://github.com/elastic/kibana/assets/7405507/287a6798-0f5e-4a7a-bf4e-811f465ef5d7">

peteharverson reviewed (approved):
Tested latest updates LGTM - fixed height for the context chart works well.

As @walterra found, there seems to be a minor issue with the bottom margin growing in height, but that shouldn't block this enhancement going in.
